### DISCUSSION TITLE
Puma worker accumulates high memory

### DISCUSSION THREAD

**Discussion by ranhorev1** [puma/puma#3063] [Q&A]
Hi!
although using puma worker killer, I am trying to figure out this memory leak \ consumption issue we have in our puma (rails application).
The reason I mention Puma, is that using different memory profilers, I keep observing that Ruby consumed normal amount of memory, but the puma worker is constantly increasing!
I am using Puma 4.3.12 Ruby 2.7.6 Rails 6.0.3.3

in order to reproduce this issue, I generated many calls to a certain endpoint in my app, and I took 3 memory dumps - Before, During and After.

As you will see in the following pastes, you will see that after the operation, the app's memory returned to its normal state, while the puma workers actually kept holding high memory!

I will appreciate any tips on how to continue investigating this issue, or known bugs (which I could not find) and a way to solve..
Thanks!

**BEFORE:**
KiB Mem :  2007248 total,   628156 free,   766956 used,   612136 buff/cache
KiB Swap:  2097148 total,  2049276 free,    47872 used.  1064852 avail Mem

Total allocated: 115227 bytes (1433 objects)
Total retained:  21036 bytes (147 objects)

allocated memory by gem
-----------------------------------
     33121  activesupport-6.0.4.7
     21687  actionpack-6.0.4.7
     14484  activerecord-6.0.4.7
     12582  var/app
      9904  ipaddr

**DURING:**
KiB Mem :  2007248 total,    65068 free,  1800424 used,   141756 buff/cache
KiB Swap:  2097148 total,  2047228 free,    49920 used.    58376 avail Mem

Total allocated: 225272583 bytes (942506 objects)
Total retained:  1732241 bytes (12035 objects)

allocated memory by gem
-----------------------------------
 106497060  maxmind-db-1.0.0
  58308032  psych
  38857594  user_agent_parser-2.7.0
   4949108  activesupport-6.0.4.7
   3967930  other
   3229962  activerecord-6.0.4.7
   2154670  rack-2.2.4
   1467383  actionpack-6.0.4.7
   1336204  activemodel-6.0.4.7
   1125325  uri

**AFTER:**
KiB Mem :  2007248 total,    73760 free,  1817688 used,   115800 buff/cache
KiB Swap:  2097148 total,  2032636 free,    64512 used.    54448 avail Mem

Total allocated: 109563 bytes (1398 objects)
Total retained:  14988 bytes (110 objects)

allocated memory by gem
-----------------------------------
     29745  activesupport-6.0.4.7
     21495  actionpack-6.0.4.7
     13452  activerecord-6.0.4.7
     12502  var/app
      9904  ipaddr
      7237  rack-2.2.4

![image](https://user-images.githubusercontent.com/94047710/213915610-391002be-89f5-4506-8550-97ae4c82aca9.png)

**Comment by nateberkopec** [comment]
If you look through our [issue tracker](https://github.com/puma/puma/issues?q=is%3Aissue+memory+leak+) you'll find a lot of suggestions and discussion on how to look for memory leaks in your app.

**Reply by ranhorev1** [reply]
@nateberkopec Thank you for your answer, i've been reading and researching extensively on memory leak \ bloat including going over the known issues (of course I could have missed something).
This post comes after adopting the different suggestions, and analyzing the memory usage with derailed benchmarks and mini-profiler.
All memory tests and reports indicate that the application does not hold any (out of the ordinary) extra memory (see "After" part), Which is why I pasted the heap analysis and was hoping for some more inputs on the issue.
